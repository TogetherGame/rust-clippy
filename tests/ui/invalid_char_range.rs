#![warn(clippy::invalid_char_range)]

const VALID: u32 = 0xD000;
const INVALID_BETWEEN: u32 = 0xDDDD;
const INVALID_LARGE_SIGNED: i64 = 0xFFFFFF;

pub fn valid() {
    let _ = char::from(97u8);
    let _ = char::from_u32(0x10FFFF);
    let _ = char::from_u32(VALID);
}

pub fn invalid() {
    let _ = char::from_u32(0xD800);
    //~^ ERROR: converting to char with out-of-range integer
    //~| NOTE: `-D clippy::invalid-char-range` implied by `-D warnings`
    let _ = unsafe { char::from_u32_unchecked(0xD800) };
    //~^ ERROR: converting to char with out-of-range integer
    let _ = char::from_u32(INVALID_BETWEEN);
    //~^ ERROR: converting to char with out-of-range integer
    let _ = char::from_u32(INVALID_BETWEEN + 1);
    //~^ ERROR: converting to char with out-of-range integer
    let _ = unsafe { char::from_u32_unchecked(INVALID_BETWEEN + 1) };
    //~^ ERROR: converting to char with out-of-range integer
    let _ = char::from_u32(INVALID_LARGE_SIGNED as u32);
    //~^ ERROR: converting to char with out-of-range integer
    let _ = char::from_u32(INVALID_LARGE_SIGNED as usize as u32);
    //~^ ERROR: converting to char with out-of-range integer
}

mod my_char {
    pub fn from_u32(u: u32) -> char {
        'a'
    }
}

pub fn skip() {
    let _ = my_char::from_u32(INVALID_BETWEEN);
}

fn get_num() -> u32 {
    100
}
// Don't lint uncertain results
pub fn uncertain(n: u32) {
    const INVALID_BETWEEN: u32 = 0xDDDD;
    let _ = char::from_u32(INVALID_BETWEEN.saturating_sub(1));

    let _ = char::from_u32(n);
    let _ = char::from_u32(10 + n);
    let _ = char::from_u32(10_u32.saturating_add(n));

    let _ = char::from_u32(get_num());
    let clo = || 0;
    let _ = char::from_u32(clo());
}

fn main() {}
