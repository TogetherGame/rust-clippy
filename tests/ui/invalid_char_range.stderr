error: converting to char with out-of-range integer
  --> $DIR/invalid_char_range.rs:14:13
   |
LL |     let _ = char::from_u32(0xD800);
   |             ^^^^^^^^^^^^^^^^^^^^^^
   |
note: this number should be within the range of [0, 0xD7FF] or [0xE000, 0x10FFFF]
  --> $DIR/invalid_char_range.rs:14:28
   |
LL |     let _ = char::from_u32(0xD800);
   |                            ^^^^^^
   = note: `-D clippy::invalid-char-range` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::invalid_char_range)]`

error: converting to char with out-of-range integer
  --> $DIR/invalid_char_range.rs:17:22
   |
LL |     let _ = unsafe { char::from_u32_unchecked(0xD800) };
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this number should be within the range of [0, 0xD7FF] or [0xE000, 0x10FFFF]
  --> $DIR/invalid_char_range.rs:17:47
   |
LL |     let _ = unsafe { char::from_u32_unchecked(0xD800) };
   |                                               ^^^^^^

error: converting to char with out-of-range integer
  --> $DIR/invalid_char_range.rs:19:13
   |
LL |     let _ = char::from_u32(INVALID_BETWEEN);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this number should be within the range of [0, 0xD7FF] or [0xE000, 0x10FFFF]
  --> $DIR/invalid_char_range.rs:19:28
   |
LL |     let _ = char::from_u32(INVALID_BETWEEN);
   |                            ^^^^^^^^^^^^^^^

error: converting to char with out-of-range integer
  --> $DIR/invalid_char_range.rs:21:13
   |
LL |     let _ = char::from_u32(INVALID_BETWEEN + 1);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this number should be within the range of [0, 0xD7FF] or [0xE000, 0x10FFFF]
  --> $DIR/invalid_char_range.rs:21:28
   |
LL |     let _ = char::from_u32(INVALID_BETWEEN + 1);
   |                            ^^^^^^^^^^^^^^^^^^^

error: converting to char with out-of-range integer
  --> $DIR/invalid_char_range.rs:23:22
   |
LL |     let _ = unsafe { char::from_u32_unchecked(INVALID_BETWEEN + 1) };
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this number should be within the range of [0, 0xD7FF] or [0xE000, 0x10FFFF]
  --> $DIR/invalid_char_range.rs:23:47
   |
LL |     let _ = unsafe { char::from_u32_unchecked(INVALID_BETWEEN + 1) };
   |                                               ^^^^^^^^^^^^^^^^^^^

error: converting to char with out-of-range integer
  --> $DIR/invalid_char_range.rs:25:13
   |
LL |     let _ = char::from_u32(INVALID_LARGE_SIGNED as u32);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this number should be within the range of [0, 0xD7FF] or [0xE000, 0x10FFFF]
  --> $DIR/invalid_char_range.rs:25:28
   |
LL |     let _ = char::from_u32(INVALID_LARGE_SIGNED as u32);
   |                            ^^^^^^^^^^^^^^^^^^^^

error: converting to char with out-of-range integer
  --> $DIR/invalid_char_range.rs:27:13
   |
LL |     let _ = char::from_u32(INVALID_LARGE_SIGNED as usize as u32);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this number should be within the range of [0, 0xD7FF] or [0xE000, 0x10FFFF]
  --> $DIR/invalid_char_range.rs:27:28
   |
LL |     let _ = char::from_u32(INVALID_LARGE_SIGNED as usize as u32);
   |                            ^^^^^^^^^^^^^^^^^^^^

error: aborting due to 7 previous errors

